File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/OSOutcomeEventsRepository.java
package com.onesignal.outcomes;

import com.onesignal.OSLogger;
import com.onesignal.OneSignalApiResponseHandler;
import com.onesignal.influence.model.OSInfluence;
import com.onesignal.outcomes.domain.OutcomeEventsService;
import com.onesignal.outcomes.model.OSOutcomeEventParams;

import java.util.List;
import java.util.Set;

abstract class OSOutcomeEventsRepository implements com.onesignal.outcomes.domain.OSOutcomeEventsRepository {

    static final String APP_ID = "app_id";
    static final String DEVICE_TYPE = "device_type";

    protected final OSLogger logger;
    private final OSOutcomeEventsCache outcomeEventsCache;
    final OutcomeEventsService outcomeEventsService;

    OSOutcomeEventsRepository(OSLogger logger, OSOutcomeEventsCache outcomeEventsCache, OutcomeEventsService outcomeEventsService) {
        this.logger = logger;
        this.outcomeEventsCache = outcomeEventsCache;
        this.outcomeEventsService = outcomeEventsService;
    }

    public abstract void requestMeasureOutcomeEvent(String appId, int deviceType, OSOutcomeEventParams event, OneSignalApiResponseHandler responseHandler);

    public List<OSOutcomeEventParams> getSavedOutcomeEvents() {
        return outcomeEventsCache.getAllEventsToSend();
    }

    public void saveOutcomeEvent(OSOutcomeEventParams event) {
        outcomeEventsCache.saveOutcomeEvent(event);
    }

    public void removeEvent(OSOutcomeEventParams outcomeEvent) {
        outcomeEventsCache.deleteOldOutcomeEvent(outcomeEvent);
    }

    public void saveUniqueOutcomeNotifications(OSOutcomeEventParams eventParams) {
        outcomeEventsCache.saveUniqueOutcomeEventParams(eventParams);
    }

    public List<OSInfluence> getNotCachedUniqueOutcome(String name, List<OSInfluence> influences) {
        List<OSInfluence> influencesNotCached =  outcomeEventsCache.getNotCachedUniqueInfluencesForOutcome(name, influences);
        logger.debug("OneSignal getNotCachedUniqueOutcome influences: " + influencesNotCached);
        return influencesNotCached;
    }

    @Override
    public Set<String> getUnattributedUniqueOutcomeEventsSent() {
        Set<String> unattributedUniqueOutcomeEvents = outcomeEventsCache.getUnattributedUniqueOutcomeEventsSentByChannel();
        logger.debug("OneSignal getUnattributedUniqueOutcomeEventsSentByChannel: " + unattributedUniqueOutcomeEvents);
        return unattributedUniqueOutcomeEvents;
    }

    @Override
    public void saveUnattributedUniqueOutcomeEventsSent(Set<String> unattributedUniqueOutcomeEvents) {
        logger.debug("OneSignal save unattributedUniqueOutcomeEvents: " + unattributedUniqueOutcomeEvents);
        outcomeEventsCache.saveUnattributedUniqueOutcomeEventsSentByChannel(unattributedUniqueOutcomeEvents);
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/data/OSOutcomeEventsRepository.kt
package com.onesignal.outcomes.data

import com.onesignal.OSLogger
import com.onesignal.OneSignalApiResponseHandler
import com.onesignal.influence.domain.OSInfluence
import com.onesignal.outcomes.domain.OSOutcomeEventParams
import com.onesignal.outcomes.domain.OSOutcomeEventsRepository

internal abstract class OSOutcomeEventsRepository(protected val logger: OSLogger,
                                                  private val outcomeEventsCache: OSOutcomeEventsCache,
                                                  val outcomeEventsService: OutcomeEventsService) : OSOutcomeEventsRepository {
    abstract override fun requestMeasureOutcomeEvent(appId: String, deviceType: Int, event: OSOutcomeEventParams, responseHandler: OneSignalApiResponseHandler)

    override fun getSavedOutcomeEvents(): List<OSOutcomeEventParams> = outcomeEventsCache.getAllEventsToSend()

    override fun saveOutcomeEvent(event: OSOutcomeEventParams) {
        outcomeEventsCache.saveOutcomeEvent(event)
    }

    override fun removeEvent(outcomeEvent: OSOutcomeEventParams) {
        outcomeEventsCache.deleteOldOutcomeEvent(outcomeEvent)
    }

    override fun saveUniqueOutcomeNotifications(eventParams: OSOutcomeEventParams) {
        outcomeEventsCache.saveUniqueOutcomeEventParams(eventParams)
    }

    override fun getNotCachedUniqueOutcome(name: String, influences: List<OSInfluence>): List<OSInfluence> {
        val influencesNotCached = outcomeEventsCache.getNotCachedUniqueInfluencesForOutcome(name, influences)
        logger.debug("OneSignal getNotCachedUniqueOutcome influences: $influencesNotCached")
        return influencesNotCached
    }

    override fun getUnattributedUniqueOutcomeEventsSent(): Set<String>? {
        val unattributedUniqueOutcomeEvents = outcomeEventsCache.unattributedUniqueOutcomeEventsSentByChannel
        logger.debug("OneSignal getUnattributedUniqueOutcomeEventsSentByChannel: $unattributedUniqueOutcomeEvents")
        return unattributedUniqueOutcomeEvents
    }

    override fun saveUnattributedUniqueOutcomeEventsSent(unattributedUniqueOutcomeEvents: Set<String>) {
        logger.debug("OneSignal save unattributedUniqueOutcomeEvents: $unattributedUniqueOutcomeEvents")
        outcomeEventsCache.saveUnattributedUniqueOutcomeEventsSentByChannel(unattributedUniqueOutcomeEvents)
    }

    override fun cleanCachedUniqueOutcomeEventNotifications(notificationTableName: String, notificationIdColumnName: String) {
        outcomeEventsCache.cleanCachedUniqueOutcomeEventNotifications(notificationTableName, notificationIdColumnName);
    }
}